packages/sdks/src/blocks/personalization-container/helpers/inlined-fns.ts

```diff
@@ -194,6 +174,65 @@ export function filterWithCustomTargeting(
   });
 }
 
+export function updateVisibilityStylesScript(
+  variants: PersonalizationContainerProps['variants'],
+  blockId: string,
+  isHydrationTarget: boolean,
+  locale?: string
+) {
+  function getCookie(name: string) {
+    const nameEQ = name + '=';
+    const ca = document.cookie.split(';');
+    for (let i = 0; i < ca.length; i++) {
+      let c = ca[i];
+      while (c.charAt(0) == ' ') c = c.substring(1, c.length);
+      if (c.indexOf(nameEQ) == 0) return c.substring(nameEQ.length, c.length);
+    }
+    return null;
+  }
+  const visibilityStylesEl = document.currentScript
+    ?.previousElementSibling as HTMLStyleElement;
+
+  if (!visibilityStylesEl) {
+    return;
+  }
+
+  if (isHydrationTarget) {
+    visibilityStylesEl.remove();
+    const currentScript = document.currentScript;
+    if (currentScript) {
+      currentScript.remove();
+    }
+  } else {
+    const attributes = JSON.parse(getCookie('builder.userAttributes') || '{}');
+    if (locale) {
+      attributes.locale = locale;
+    }
+    const winningVariantIndex = variants?.findIndex(function (variant) {
+      return (window as any).filterWithCustomTargeting(
+        attributes,
+        variant.query,
+        variant.startDate,
+        variant.endDate
+      );
+    });
+
```

Comment:
one idea here to optimize finding the winner again and again is to set a cookie - something like `builder.pblock.id` to the winning index or -1 and then pick that instead in the personalization script? but it will be an extra cookie addition for each block

Reply:
@samijaber